src/Composer/Util/Http/ProxyItem.php

```
@@ -0,0 +1,151 @@
+<?php declare(strict_types=1);
+
+/*
+ * This file is part of Composer.
+ *
+ * (c) Nils Adermann <naderman@naderman.de>
+ *     Jordi Boggiano <j.boggiano@seld.be>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Composer\Util\Http;
+
+/**
+ * @internal
+ * @author John Stevenson <john-stevenson@blueyonder.co.uk>
+ *
+ * @phpstan-type contextOptions array{http: array{proxy: string, header?: string, request_fulluri?: bool}}
+ */
+class ProxyItem
+{
+    /** @var non-empty-string */
+    private $url;
+    /** @var non-empty-string */
+    private $safeUrl;
+    /** @var ?non-empty-string */
+    private $curlAuth;
+    /** @var string */
+    private $optionsProxy;
+    /** @var ?non-empty-string */
+    private $optionsAuth;
+
+    public function __construct(string $proxy, string $envName)
+    {
+        if (strpbrk($proxy, "\r\n\t") !== false || !$this->checkData($proxy)) {
+            throw new \RuntimeException(sprintf('unsupported `%s` syntax', $envName));
+        }
```

Maybe move this check to ProxyHandler::getProxyEnv, then it can simply return $env and no name, and this class doesn't need $envName either.

---

Ah no sorry you kinda need that as checkData can return false too. But I'd still rather move this whole check in checkData, or just move checkData body to be the constructor body, so you can throw proper exception message withw hat is unsupported instead of just a generic throw when checkData returns false.